SeQuant/core/eval_expr.hpp

```diff
@@ -1,190 +1,50 @@
- *
- * @author Bimal Gaudel
- * @version 03 Dec, 2020
- */
-class EvalExpr {
- public:
-  using hash_t = size_t;
+enum class EvalOp { Id, Sum, Prod };
+
+enum class ResultType { Tensor, Constant };
 
-  EvalExpr() = delete;
+class EvalExpr {
+ private:
+  EvalOp op_type_;
+  ResultType result_type_;
+  size_t hash_value_;
+  size_t id_;
+  ExprPtr expr_;
+  bool tot_;
 
-  /**
-   * Construct EvalExpr that goes into the leaf nodes
-   * of the binary evaluation tree.
-   */
-  explicit EvalExpr(const Tensor&) noexcept;
+  static size_t global_id_;
 
-  /**
-   * Construct EvalExpr that goes into the internal nodes
-   * of the binary evaluation tree.
-   */
-  EvalExpr(const EvalExpr&, const EvalExpr&, EvalOp op) noexcept;
+ public:
+  explicit EvalExpr(Tensor const& tnsr);
 
-#ifdef DEBUG_EVAL_EXPR
-  EvalExpr(EvalExpr&&) noexcept;
+  explicit EvalExpr(Constant const& c);
 
-  EvalExpr& operator=(EvalExpr&&) noexcept;
+  EvalExpr(EvalExpr const& left, EvalExpr const& right, EvalOp op);
 
-  EvalExpr(EvalExpr const&) noexcept;
+  [[nodiscard]] EvalOp op_type() const noexcept;
 
-  EvalExpr& operator=(EvalExpr const&) noexcept;
+  [[nodiscard]] ResultType result_type() const noexcept;
 
-  ~EvalExpr() noexcept;
-#endif
+  [[nodiscard]] size_t hash_value() const noexcept;
 
-  /**
-   * The unique id of this object.
-   */
   [[nodiscard]] size_t id() const noexcept;
 
-  /**
-   * The store Tensor's label plus the id.
-   */
-  [[nodiscard]] std::string label(std::string_view sep = "_") const noexcept;
-
-  /**
-   * Operation type of the object.
-   */
-  [[nodiscard]] EvalOp op() const noexcept;
-
-  /**
-   * Hash value of the object.
-   */
-  [[nodiscard]] hash_t hash_value() const noexcept;
-
-  /**
-   * Tensor expression stored by the object.
-   */
-  [[nodiscard]] const Tensor& tensor() const noexcept;
-
-  /** Factor to scale tensor by. */
-  [[nodiscard]] const Constant& scalar() const noexcept;
-
-  /** Scale the scalar prefactor by @c fac. */
-  template <typename T = std::complex<double>>
-  EvalExpr& operator*=(T fac) noexcept {
-    scalar_ *= Constant{std::move(fac)};
-    return *this;
-  }
-
-  /** Set the scalar prefactor to @c fac. */
-  template <typename T = std::complex<double>>
-  void scale(T fac) noexcept {
-    scalar_ = Constant{std::move(fac)};
-  }
-
-  friend inline bool operator==(const EvalExpr& lhs,
-                                const EvalExpr& rhs) noexcept {
-    return lhs.hash_value() == rhs.hash_value() &&
-           (lhs.tensor().to_latex() == rhs.tensor().to_latex());
-  }
+  [[nodiscard]] ExprPtr expr() const noexcept;
 
- private:
-  using index_container_type = container::svector<Index>;
-  using braket_type = std::pair<index_container_type, index_container_type>;
+  [[nodiscard]] bool tot() const noexcept;
 
-  size_t id_;
+  [[nodiscard]] std::wstring to_latex() const noexcept;
+
+  [[nodiscard]] sequant::Tensor const& as_tensor() const noexcept(false);
```

Comment:
```suggestion
  [[nodiscard]] sequant::Tensor const& as_tensor() const;
```
There is no point in adding `noexcept(false)` to a regular function. The only case where this can make sense is with destructors as these are `noexcept` by default. See also https://stackoverflow.com/q/16244313